### DISCUSSION TITLE
Code Review: Monad and run time error

### DISCUSSION THREAD

**Discussion by cd155** [JacquesCarette/Drasil#3297] [Show and tell]
In the following code, it will have run time error because `isBalancedWithHeight'` could return a `Nothing`, I did not tell the program what to do if  `isBalancedWithHeight' l` or `isBalancedWithHeight' r` is `Nothing`. (the comment part)

```haskell
isBalancedWithHeight' :: BiTree a -> Maybe Int
isBalancedWithHeight' Empty' = Just 0
isBalancedWithHeight' (Node' _ (l,r))
    -- | isBalancedWithHeight l == Nothing || isBalancedWithHeight r == Nothing = 
    --     Nothing
    | abs(lh - rh) <= 1 = Just (max lh rh + 1)
    | otherwise = Nothing
    where Just lh = isBalancedWithHeight' l
          Just rh = isBalancedWithHeight' r
```

I was puzzled why the following code wouldn't have a run time error as the previous code does
```haskell
isBalancedWithHeight :: BiTree a -> Maybe Int
isBalancedWithHeight Empty' = Just 0
isBalancedWithHeight (Node' _ (l,r)) = do
  lh <- isBalancedWithHeight l
  rh <- isBalancedWithHeight r
  if abs (lh-rh) <= 1
    then Just $ max lh rh + 1
    else Nothing
```

After enabling Haskell Language Server in vscode, `lh` and `rh` be automatically inferred as `Int` rather than `Mabye Int`. Can anyone tell me why and how this code handle when `isBalancedWithHeight l` or `isBalancedWithHeight r` return `Nothing`?

**Comment by cd155** [comment]
The following is responded by ChatGPT

The reason why `lh` and `rh` are inferred as Int rather than `Maybe Int` is because they are extracted using the `<-` operator inside a `do` block, which automatically unwraps the `Just` value inside the `Maybe` monad.

Regarding the handling of `Nothing` values, in the `do` block, if `isBalancedWithHeight l` or `isBalancedWithHeight r` returns `Nothing`, the entire computation will short-circuit and return `Nothing`, indicating that the tree is not balanced. This is because in a `do` block, if any step in the sequence returns `Nothing`, the entire computation returns `Nothing`. Therefore, you do not need to handle the case explicitly like in the commented out section of the first code snippet.

**Reply by smiths** [reply]
I'm very far from an expert in Haskell, but this looks like a reasonable response from ChatGPT.  @cd155 you gave it the above code snippets and asked your question?  Maybe we can replace some TAs with AI in our courses.  :smile:  Of course I'd rather increase TA hours, but having another option for students to learn is a good thing.

**Reply by cd155** [reply]
I am satisfied with the answer that ChatGPT gave, but I feel something is still missing. For example, the chat didn't say why the `<-` can automatically unwrap the value in the 'do' block. Other than that, the response is quite impressive.

**Reply by cd155** [reply]
@smiths I also hear some people use ChatGPT to do their homework. I know most institutions use plagiarism software. It seems a challenge that artifacts work is quite similar to human creation. It also raises a question of how we distinguish artifacts and human creation.

**Reply by smiths** [reply]
Yes, I think ChatGPT is becoming an issue for homework.  I'm lucky to be teaching the capstone project right now where it isn't a worry.  However, in my second year design course in the past many of my assignment questions could have been answered by ChatGPT.  If I was working on a course like that now, I would have to spend time checking my questions with ChatGPT.  I know the Faculty is investigating options for detecting academic dishonesty using ChatGPT.

**Reply by JacquesCarette** [reply]
ChatGPT has a correct answer here, even though it's phrased in a way that you'd expect a beginner to use and not an experienced Haskell programmer.

The `<-` isn't "unwrapping" the `Int`, it is using the definition of what happens in the `Maybe` monad, specifically the meaning of bind (aka `>>=`). The second paragraph alludes to that, in a roundabout way.

I was speaking with a prof yesterday who caught multiple (Master's !!!) students in his course using ChatGPT for handing in answers to some essay questions. It's a real problem, and it will get worse before it gets better.
